### ISSUE TITLE
Clang wrongly reports duplicate member error on using recursive template

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#38611]
|  |  |
| --- | --- |
| Bugzilla Link | [39263](https://llvm.org/bz39263) |
| Version | 7.0 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@efriedma-quic,@zygoloid |

## Extended Description 
I am using Clang 7.0 on Windows 10, but could reproduce the issue on Compiler Explorer using different compiler versions.

The following code compiles fine on GCC (tested with 4.2 (!) and 4.8 and other versions available on Compiler Explorer), and on MSVC 19 (again on CE).

But yields error on every clang I tried, both on my local machine as well as on Compiler Explorer.

The code requires boost (tested with 1.52 and newer)

Code:

```cpp
#include <boost/mpl/integral_c.hpp>

struct r_base {};

typedef long long r_value_t;

template<r_value_t N, r_value_t D = 1>
struct r : r_base
{
    static const r_value_t n = N;
    static const r_value_t d = D;
};

template<r_value_t X>
struct a : boost::mpl::integral_c<r_value_t, (X < 0) ? -X : X> {};

template<r_value_t U, r_value_t V>
struct g : g<V, U % V> {};

template<r_value_t U>
struct g<U, 0> : a<U> {};

template<typename T>
struct r_r
{
    static const r_value_t n = a<T::n>::value;
    static const r_value_t d = a<T::d>::value; // error happens here: d is duplicate member
    static const r_value_t g = g<n, d>::value;
};
```
Clang reports the errors:
```
<source>:29:37: error: duplicate member 'd'
    static const r_value_t g = g<n, d>::value;
                                    ^
<source>:28:28: note: previous declaration is here
    static const r_value_t d = a<T::d>::value;
                           ^
<source>:29:38: error: expected ';' at end of declaration list
    static const r_value_t g = g<n, d>::value;
                                     ^
                                     ;
```

Thank you in advance.

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [39263](https://llvm.org/bz39263) |
| Version | 7.0 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@efriedma-quic,@zygoloid |

## Extended Description 
I am using Clang 7.0 on Windows 10, but could reproduce the issue on Compiler Explorer using different compiler versions.

The following code compiles fine on GCC (tested with 4.2 (!) and 4.8 and other versions available on Compiler Explorer), and on MSVC 19 (again on CE).

But yields error on every clang I tried, both on my local machine as well as on Compiler Explorer.

The code requires boost (tested with 1.52 and newer)

Code:

```cpp
#include &lt;boost/mpl/integral_c.hpp&gt;

struct r_base {};

typedef long long r_value_t;

template&lt;r_value_t N, r_value_t D = 1&gt;
struct r : r_base
{
    static const r_value_t n = N;
    static const r_value_t d = D;
};

template&lt;r_value_t X&gt;
struct a : boost::mpl::integral_c&lt;r_value_t, (X &lt; 0) ? -X : X&gt; {};

template&lt;r_value_t U, r_value_t V&gt;
struct g : g&lt;V, U % V&gt; {};

template&lt;r_value_t U&gt;
struct g&lt;U, 0&gt; : a&lt;U&gt; {};

template&lt;typename T&gt;
struct r_r
{
    static const r_value_t n = a&lt;T::n&gt;::value;
    static const r_value_t d = a&lt;T::d&gt;::value; // error happens here: d is duplicate member
    static const r_value_t g = g&lt;n, d&gt;::value;
};
```
Clang reports the errors:
```
&lt;source&gt;:29:37: error: duplicate member 'd'
    static const r_value_t g = g&lt;n, d&gt;::value;
                                    ^
&lt;source&gt;:28:28: note: previous declaration is here
    static const r_value_t d = a&lt;T::d&gt;::value;
                           ^
&lt;source&gt;:29:38: error: expected ';' at end of declaration list
    static const r_value_t g = g&lt;n, d&gt;::value;
                                     ^
                                     ;
```

Thank you in advance.
</details>

**Comment by shafik** [comment]
Fixed a while ago: https://godbolt.org/z/Yovo7sW5q
